flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java

```diff
@@ -147,35 +146,34 @@ public void stop() {
-                                    ExceptionUtils.findThrowable(
-                                            t, UnsuccessfulExecutionException.class);
-                            if (maybeException.isPresent()) {
-                                final ApplicationStatus applicationStatus =
-                                        maybeException.get().getStatus();
-                                if (applicationStatus == ApplicationStatus.CANCELED
-                                        || applicationStatus == ApplicationStatus.FAILED) {
-                                    LOG.info("Application {}: ", applicationStatus, t);
-                                    return dispatcherGateway.shutDownCluster(applicationStatus);
-                                }
-                            }
-
-                            LOG.warn("Application failed unexpectedly: ", t);
-                            this.errorHandler.onFatalError(
-                                    new FlinkException("Application failed unexpectedly.", t));
-
-                            return FutureUtils.<Acknowledge>completedExceptionally(t);
-                        })
-                .thenCompose(Function.identity());
+        final CompletableFuture<Acknowledge> shutdownFuture =
+                applicationCompletionFuture
+                        .handle(
+                                (ignored, t) -> {
+                                    if (t == null) {
+                                        LOG.info("Application completed SUCCESSFULLY");
+                                        return dispatcherGateway.shutDownCluster(
+                                                ApplicationStatus.SUCCEEDED);
+                                    }
+                                    final Optional<UnsuccessfulExecutionException> maybeException =
+                                            ExceptionUtils.findThrowable(
+                                                    t, UnsuccessfulExecutionException.class);
+                                    if (maybeException.isPresent()) {
+                                        final ApplicationStatus applicationStatus =
+                                                maybeException.get().getStatus();
+                                        if (applicationStatus == ApplicationStatus.CANCELED
+                                                || applicationStatus == ApplicationStatus.FAILED) {
+                                            LOG.info("Application {}: ", applicationStatus, t);
+                                            return dispatcherGateway.shutDownCluster(
+                                                    applicationStatus);
+                                        }
+                                    }
+                                    LOG.warn("Application failed unexpectedly: ", t);
+                                    return dispatcherGateway.shutDownClusterExceptionally(t);
```

Comment:
Why is it better to shut down the cluster exceptionally than failing hard here?

Reply:
I actually think that this might lead to an undesired behaviour. In case of an unexpected failure we would like to restart the process. However, if we shut the cluster down, then it won't get restarted (e.g. on Yarn) because we deregister the application.

Reply:
You're right. Another approach could be giving up leadership / restarting dispatcher, but this could be tricky to do 🤔 Not sure if that would actually improve anything, probably not worth the effort.